File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/academy/QuestionnaireViewModel.kt
package br.church.paz.android.ui.features.academy

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.church.paz.shared.domain.repository.CourseRepository
import br.church.paz.shared.domain.repository.QuestionnaireAnswerInput
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class QuestionnaireViewModel(
    private val courseId: String,
    private val courseRepository: CourseRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(QuestionnaireUiState())
    val uiState: StateFlow<QuestionnaireUiState> = _uiState.asStateFlow()

    private val _effect = Channel<QuestionnaireEffect>(Channel.BUFFERED)
    val effect = _effect.receiveAsFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            runCatching { courseRepository.getQuestionnaire(courseId) }
                .onSuccess { questionnaire ->
                    _uiState.update { it.copy(isLoading = false, questionnaire = questionnaire) }
                }.onFailure { e ->
                    _uiState.update { it.copy(isLoading = false, error = e.message) }
                }
        }
    }

    fun onSelectSingleOption(
        questionId: String,
        optionId: String,
    ) {
        _uiState.update {
            it.copy(selectedOptionIds = it.selectedOptionIds + (questionId to setOf(optionId)))
        }
    }

    fun onToggleMultiOption(
        questionId: String,
        optionId: String,
    ) {
        _uiState.update { state ->
            val current = state.selectedOptionIds[questionId] ?: emptySet()
            val updated = if (optionId in current) current - optionId else current + optionId
            state.copy(selectedOptionIds = state.selectedOptionIds + (questionId to updated))
        }
    }

    fun onFreeTextChanged(
        questionId: String,
        text: String,
    ) {
        _uiState.update { it.copy(freeTextAnswers = it.freeTextAnswers + (questionId to text)) }
    }

    fun onNextStep() {
        val questions =
            _uiState.value.questionnaire
                ?.questions
                .orEmpty()
        _uiState.update { it.copy(stepIndex = (it.stepIndex + 1).coerceAtMost(questions.size - 1)) }
    }

    fun onPreviousStep() {
        _uiState.update { it.copy(stepIndex = (it.stepIndex - 1).coerceAtLeast(0)) }
    }

    fun onBack() {
        viewModelScope.launch { _effect.send(QuestionnaireEffect.NavigateBack) }
    }

    fun onSubmit() {
        val questionnaire = _uiState.value.questionnaire ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmitting = true, submitError = null) }
            val answers =
                questionnaire.questions.map { question ->
                    QuestionnaireAnswerInput(
                        questionId = question.id,
                        optionIds =
                            _uiState.value.selectedOptionIds[question.id]
                                ?.toList()
                                .orEmpty(),
                        text = _uiState.value.freeTextAnswers[question.id],
                    )
                }
            runCatching { courseRepository.submitQuestionnaire(courseId, answers) }
                .onSuccess { result ->
                    _uiState.update { it.copy(isSubmitting = false, result = result) }
                }.onFailure { e ->
                    _uiState.update { it.copy(isSubmitting = false, submitError = e.message) }
                }
        }
    }
}

File: kmp-mobile/ios/PazChurch/Features/Academy/QuestionnaireViewModel.swift
import Observation
import Shared
import SwiftUI

@MainActor
@Observable
class QuestionnaireViewModel {
    var isLoading = true
    var questionnaire: Questionnaire?
    var error: String?
    var stepIndex = 0
    var selectedOptionIds: [String: Set<String>] = [:]
    var freeTextAnswers: [String: String] = [:]
    var isSubmitting = false
    var submitError: String?
    var result: QuestionnaireResult?

    private let courseId: String
    private let courseRepository: CourseRepository

    var progress: Double {
        guard let count = questionnaire?.questions.count, count > 0 else { return 0 }
        return Double(stepIndex + 1) / Double(count)
    }

    var isLastStep: Bool {
        guard let count = questionnaire?.questions.count else { return true }
        return stepIndex == count - 1
    }

    init(courseId: String, courseRepository: CourseRepository) {
        self.courseId = courseId
        self.courseRepository = courseRepository
    }

    func load() async {
        isLoading = true
        error = nil
        do {
            questionnaire = try await courseRepository.getQuestionnaire(courseId: courseId)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func selectSingle(questionId: String, optionId: String) {
        selectedOptionIds[questionId] = [optionId]
    }

    func toggleMulti(questionId: String, optionId: String) {
        var current = selectedOptionIds[questionId] ?? []
        if current.contains(optionId) {
            current.remove(optionId)
        } else {
            current.insert(optionId)
        }
        selectedOptionIds[questionId] = current
    }

    func setFreeText(questionId: String, text: String) {
        freeTextAnswers[questionId] = text
    }

    func nextStep() {
        guard let count = questionnaire?.questions.count else { return }
        stepIndex = min(stepIndex + 1, count - 1)
    }

    func previousStep() {
        stepIndex = max(stepIndex - 1, 0)
    }

    func onSubmit() {
        guard let questionnaire else { return }
        isSubmitting = true
        submitError = nil
        Task {
            let answers = questionnaire.questions.map { question in
                QuestionnaireAnswerInput(
                    questionId: question.id,
                    optionIds: Array(selectedOptionIds[question.id] ?? []),
                    text: freeTextAnswers[question.id]
                )
            }
            do {
                result = try await courseRepository.submitQuestionnaire(courseId: courseId, answers: answers)
            } catch {
                submitError = error.localizedDescription
            }
            isSubmitting = false
        }
    }
}
